cats/optimise_starttime.py

```diff
@@ -0,0 +1,91 @@
+            totalsec = 0 # in seconds
+            for window in filtered_forecast:
+                # find the true duration of the window [timedelta]
+                # (this is needed because first and last windows won't be a full 30min, or whatever the window size is)
+                true_window = min(endtime, window.end) - max(starttime, window.start)
+                CIsec += window.value * true_window.total_seconds()
+                totalsec += true_window.total_seconds()
+
+            return CIsec/totalsec
+
+        # TODO include again trapezoidal integration as an option
+
+    def _get_all_averageCIs(self, duration):
+        '''
+        Calculate all the average CIs for all possible start times.
+        The first possible time is now, the last possible time is "End of forecast - job duration".
+        :param duration: [timedelta]
+        :return: [list[CarbonIntensityEstimate]] All possible average CIs
+        '''
+        currenttime = datetime.now(timezone.utc)
+        end_of_forecast = self.CI_forecast[-1].end
+
+        all_averageCIs = []
+
+        # What if we start the job now
+        endtime = currenttime + duration
+        # TODO add the option to start jobs planned to run past end of forecast (i.e. with unknown CI for some of it)
+        if endtime > end_of_forecast:
+            raise ValueError(f"Job is longer than total forecasted carbon intensities (duration={duration.total_seconds()/3600:.1f} hours but only {(end_of_forecast - self.CI_forecast[0].start).total_seconds()/3600:.1f} hours available)")
+
+        all_averageCIs.append(CarbonIntensityEstimate(
+            start=currenttime,
+            end=endtime,
+            value=self._calculate_averageCI_over_runtime(currenttime, endtime),
+        ))
+
+        # Go through all the possible starttimes and calculate average CIs
+        for window in self.CI_forecast:
+            starttime = window.start
+            endtime = window.start + duration
+
+            # It needs to be (1) a starttime in the future and (2) leave enough time to finish before the end of the forecast
+            if starttime > currenttime and endtime <= end_of_forecast:
+                all_averageCIs.append(CarbonIntensityEstimate(
+                    start=starttime,
+                    end=endtime,
+                    value=self._calculate_averageCI_over_runtime(starttime, endtime),
+                ))
+
+        return all_averageCIs
+
+    def get_starttime(self, duration):
+        '''
+        Find best possible starttime to minimise carbon intensities
+        :param duration: [timedelta]
+        :return: [tuple(CarbonIntensityEstimate, list[CarbonIntensityEstimate])]
+        '''
+        all_averageCIs = self._get_all_averageCIs(duration)
+
+        averageCI_now = all_averageCIs[0]
+        # Find the minimum value
+        all_averageCIs_sorted = sorted(all_averageCIs)
+
+        return all_averageCIs_sorted[0], averageCI_now, all_averageCIs_sorted
```

Comment:
Here you're sorting a list and selecting the first element. Why not using the builtin `min()` directly?

Reply:
And the worse one a bit later, but yes happy to use min and max instead for now (sorted list may be useful in the future if we need to propose more than one time...but topic for another day.

Reply:
Changed in commit [99c88fa](https://github.com/GreenScheduler/cats/pull/43/commits/99c88fafc750066ab1b4c90c6ce6230c70501c1d)